### PR TITLE
Proposal: config data volume source

### PR DESCRIPTION
**Pull request by pmorie**
Forked from #6477, let's discuss any amendments needed to the volume here.

@tmrts @bgrant0607 @thockin @smarterclayton

### CODE REVIEW

**Comment by pmorie** [comment]
@thockin @bgrant0607 Did we wind up answering all of our questions about this?  Brian, I believe you didn't want to do templates or rolling-updates at this point.  Maybe that leaves the sentinel file -- shall we define that here?

cc @tmrts

**Comment by bgrant0607** [comment]
Did anything change between the last version of #6477 and this? Looks reasonable to me, and I believe it addresses @thockin's desire to be able to specify the path explicitly.

**Comment by tmrts** [comment]
@bgrant0607 it's the same version cut from the #6477. No modifications to it yet. We were discussing adding a sentinel file to be used as an `inotify` mechanism for clients when `ConfigData` is modified.

**Comment by davidopp** [comment]
Just wondering, who is going to implement this? Is it reasonable to assume this will be finished in the next ~6 weeks? (Not the proposal, but the implementation :) )

**Comment by tmrts** [comment]
@davidopp I've discussed with @pmorie the implementation PRs. The foundation PR is #6245 . At the moment I have limited bandwidth, so we've decided to split it 2 for me and 4 for him. @pmorie can probably give a more accurate timeline, but we have every intention to finish it before 1.2.

Also the sooner we can reach to a consensus on the volume proposal the better.
@thockin can you summarize your thoughts here?

**Comment by pmorie** [comment]
@davidopp It's one of my personal tasks to deliver this for 1.2

File: docs/proposals/config_data.md [hunk 1]
```diff
@@ -173,7 +173,36 @@ type ConfigDataSelector struct {
...
+
+type VolumeSource struct {
+  // other fields omitted
+  ConfigData *ConfigDataVolumeSource `json:"configData,omitempty"`
+}
+
+// ConfigDataVolumeSource represents a volume that holds configuration data
+type ConfigDataVolumeSource struct {
+  // A list of config data keys to project into the volume in files
+  Files []ConfigDataVolumeFile `json:"files"`
```
Comment (thockin) [inline docs/proposals/config_data.md]: DownwardAPI says "Items".  It probably should have said "Files", but it doesn't.  Consistency is more important than correctness.  File a bug to fix this for v2 api (seriously :)

**Comment by thockin** [comment]
This seems necessary but not sufficient.  Is this going to solve most use cases?  One file per datum is not a format used by any software I know.

How would I render a while ConfigData into files - do I need to call out every key?

Can I render it as a single file of key=value lines?  How about as JSON?  I'm not at all sure what is actually needed, so these are pure speculation.

How do I render a template file with a ConfigData's keys being substituted in?  That seems like the most common need.

**Comment by tmrts** [comment]
@thockin as shown in [config_data.md](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/config_data.md), key/value fields can be consumed as either filename/filecontents or envkey/envvar. This covers a good amount of use cases. 

However, I fully agree with you that users should be able to utilize templating with ``ConfigData`` at the moment there isn't any templating. But, I also believe it is not practical to deliver templating for ``1.2``. The specifications for templating are not concrete yet and the discussions usually take a good amount of time. 

Also, scope of templating is bigger than ``ConfigData`` since there will be a template with fields and fields can be filled with different data acquired from the API server (e.g. metadata.labels, configdata.keys, etc.).

I propose implementing ``ConfigData`` since it is not unusable by any means and integrate templating later on.

File: docs/proposals/config_data.md (hunk 1 above)
Comment (pmorie) [inline docs/proposals/config_data.md]: Created #17859

**Comment by pmorie** [comment]
@thockin I would like to be able to implement something that has parity with Secret (with the known flaws of Secret fixed) first, before we talk about template rendering.  Do you feel like that's an acceptable first step and enough scope for this proposal?  Templating touches secrets and the downward API volumes as well.  Should we file a separate issue for that?

**Comment by bgrant0607** [comment]
@thockin The way I expect to use this is to put a whole config file into each value.

**Comment by thockin** [comment]
The way I want to use this is with a datum per value :)

File: docs/proposals/config_data.md (hunk 1 above)
Comment (thockin) [inline docs/proposals/config_data.md]: This should be "Items" for v1

Comment (thockin) [inline docs/proposals/config_data.md]: oh hell, now I don't know.  A sensible name now but inconsistent vs consistent but to be changed later.  I can go either way.

**Comment by bgrant0607** [comment]
@thockin One datum per value makes sense for values plumbed into env vars. Why is that desirable/necessary for values composed into files?

It makes the most sense to me for application configuration generation to be a pre-deployment step.

**Comment by thockin** [comment]
I think you're right probably right for lots of cases, though I suspect it
will become more granular (more key-value-like :) as people get more
"native" with it.

Or maybe I am wrong - that's just how I thought to use it.

**Comment by davidopp** [comment]
I haven't been following this closely -- does this mean the design for the first version of config data is now finished?

**Comment by tmrts** [comment]
@davidopp That is correct. The design is now merged. The other PR (main proposal) was #6477.

**Comment by bgrant0607** [comment]
@davidopp Pretty much.
